python/inference/pytorch/make_avg_pow_net.py

```diff
@@ -0,0 +1,130 @@
+(using PyTorch) to be deployed on the AIR-T. An extremely simple network is created
+that calculates the average of the instantaneous power for an input buffer. The output
+of the application is a saved onnx file that may be optimized using NVIDIA's TensorRT
+tool to create a plan file. That plan file is deployable on the AIR-T.
+
+Users may modify the layers in the NeuralNetwork class for the network that is best
+suited for that application. Note that any layer that is used must be supported by
+TensorRT so that it can be deployed on the AIR-T. See the AirStack API documentation
+for more information on supported layers:
+http://docs.deepwavedigital.com/AirStack/airstack.html
+"""
+
+import numpy as np
+import torch
+
+
+# Top-level neural network settings
+CPLX_SAMPLES_PER_INFER = 2048  # This should be half input_len from the neural network
+ONNX_FILE_NAME = 'avg_pow_net.onnx'  # File name to save network
+INPUT_NODE_NAME = 'input_buffer'  # User defined name of input node
+OUTPUT_NODE_NAME = 'output_buffer'  # User defined name of output node
+ONNX_VERSION = 10  # the ONNX version to export the model to
+
+
+class NeuralNetwork(torch.nn.Module):
+    """ Simple network that computes the average power of an input signal
+
+    The method does this by squaring each sample, summing (I^2 + Q^2), and dividing by
+    the number of samples. Since it is assumed that the input is interleaved I/Q, the
+    divisor is input_len / 2.
+    
+    Attributes:
+        norm_matrix: pytorch tensor for normalization factor
+
+    Usage example:
+        network = NeuralNetwork(4096)
+        result = network.forward(data):
+    """
+    def __init__(self, input_len):
+        """Initialization class for method that defines normalization factor
+
+        Args:
+            input_len: int,  length of the input layer (real samples)
+
+        Returns:
+            An instance of the NeuralNetwork class:
+        """
+        super().__init__()  # Initialize superclass
+        norm_factor = 1.0 / (float(input_len) / 2.0)
+        norm_matrix = np.ones((input_len, 1)) * norm_factor
+        # Convert normalization matrix to a tensor
+        self.norm_matrix = torch.tensor(norm_matrix, dtype=torch.float32)
+        
+    def forward(self, batch_x):
+        """Forward propagation of network
+
+        Args:
+            batch_x: tensor, interleaved I and Q, i.e., <I0, Q0, I1, Q1,...,In, Qn>
+                     batch_x is of size (batch_size, input_len)
+
+        Returns:
+            output: average power tensor, i.e., avg(sig.real^2 + sig.imag^2)
+        """
+        # Create layer that squares each sample independently
+        layer1 = torch.pow(batch_x, 2.0)
+        # Create layer2 that uses a matrix multiplication to multiply each squared
+        # sample by a normalization factor then sums all samples to compute the average
+        # of layer1.
+        layer2 = torch.mm(layer1, self.norm_matrix)
```

Comment:
The way these layers are created differs from what we have in our internal test scripts. Looking at the API calls, these should result in the same network.

Reply:
Yes, I thought the TensorFlow and PyTorch examples were more consistent with each other this way. The functionality is the same using `torch.mm(layer1, self.norm_matrix)` is the same as `layer1.mm(self.norm_matrix)`.